test/workbox-precaching/node/controllers/test-PrecacheController.mjs

```diff
@@ -398,5 +401,90 @@ describe(`[workbox-precaching] PrecacheController`, function() {
         expect(cachedResponse).to.exist;
       }
     });
+
+    it('should only precache assets that have changed', async function() {
+      // Prevent logs in the mocha output
+      sandbox.stub(logger, 'warn');
+      sandbox.stub(logger, 'debug');
+      const logStub = sandbox.stub(logger, 'log');
+
+      /*
+      First precache some entries
+      */
+      const PrecacheController = (await import(PRECACHE_MANAGER_PATH)).default;
+      const precacheControllerOne = new PrecacheController();
+      const cacheListOne = [
+        '/index.1234.html',
+        {url: '/example.1234.css'},
+        {url: '/scripts/index.js', revision: '1234'},
+        {url: '/scripts/stress.js?test=search&foo=bar', revision: '1234'},
+      ];
+      precacheControllerOne.addToCacheList(cacheListOne);
+
+      // Reset as addToCacheList will log.
+      logStub.reset();
+
+      const updateInfo = await precacheControllerOne.install();
+      expect(updateInfo.updatedEntries.length).to.equal(cacheListOne.length);
+      expect(updateInfo.notUpdatedEntries.length).to.equal(0);
+
+      const cache = await caches.open(cacheNameProvider.getPrecacheName());
+      const keysOne = await cache.keys();
+      expect(keysOne.length).to.equal(cacheListOne.length);
+
+      for (let i = 0; i < cacheListOne.length; i++) {
+        // We don't cache bust requests where the revision
+        // is in the URL
+        let inputUrl = cacheListOne[i];
+        if (cacheListOne[i].url) {
+          inputUrl = cacheListOne[i].url;
+        }
+
+        let cachedResponse = await cache.match(inputUrl);
+        expect(cachedResponse).to.exist;
+      }
+
+      // Make sure we print some debug info.
+      expect(logStub.callCount).to.be.gt(0);
+
+      /*
+      THEN precache the same URLs but two with different revisions
+      */
+      const precacheControllerTwo = new PrecacheController();
+      const cacheListTwo = [
+        '/index.4321.html',
+        {url: '/example.1234.css'},
+        {url: '/scripts/index.js', revision: '1234'},
+        {url: '/scripts/stress.js?test=search&foo=bar', revision: '4321'},
+      ];
+      precacheControllerTwo.addToCacheList(cacheListTwo);
+
+      // Reset as addToCacheList will log.
+      logStub.reset();
+
+      const updateInfoTwo = await precacheControllerTwo.install();
+      expect(updateInfoTwo.updatedEntries.length).to.equal(2);
+      expect(updateInfoTwo.notUpdatedEntries.length).to.equal(2);
+
+      const keysTwo = await cache.keys();
+      // +1 allows cache keys 'index.1234.html' and 'index.4321.html'
```

Comment:
I don't follow what this comment means—is this test potentially flaky, depending on whether `activate` has fired?

Reply:
The test is isn't flaky.

This test is to check install() method behavior, not activate() method behavior.

So at this stage both 'index.1234.html' and 'index.4321.html' would exist in the cache. 'index.1234.html' from the previous precache (which may still be used by current clients) and 'index.4321.html' from the current precache (which won't be used by anything until activate event).

The activate method (Needs to be added after this PR) will perform this cleanup.